prv_accountant/composers.py

```diff
@@ -5,30 +5,58 @@
 
 from abc import ABC, abstractmethod
 from scipy.fft import rfft, irfft
+from typing import Sequence
 
 from .discrete_privacy_random_variable import DiscretePrivacyRandomVariable
 
 
 class Composer(ABC):
+    def __init__(self, prvs: Sequence[DiscretePrivacyRandomVariable]) -> None:
+        """
+        Abstract base class for a composing mechanism.
+
+        :param Sequence[DiscretePrivacyRandomVariable] prvs: Sequence of discrete PRVs to compose
+        """
+        self.prvs = prvs
+
     @abstractmethod
-    def compute_composition(self, num_compositions: int) -> DiscretePrivacyRandomVariable:
+    def compute_composition(self, num_compositions: Sequence[int]) -> DiscretePrivacyRandomVariable:
+        """
+        Abstract method to compute the composition of PRVs
+
+        :param Sequence[int] num_composition: The number of composition for each PRV with itself.
+                                              The length of this sequence needs to match `self.prvs`.
+        """
         pass
 
 
 class Fourier(Composer):
-    def __init__(self, prv: DiscretePrivacyRandomVariable) -> None:
+    def __init__(self, prvs: Sequence[DiscretePrivacyRandomVariable]) -> None:
         """
         Compute the composition of the PRVs using convolutions in Fourier space
 
-        :param DiscretePrivacyRandomVariable prv: PRV to compose
+        :param prvs: PRVs to compose.
+                     The Fourier comopser only handles homogeneous composition and therefore prvs lengths of 1.
         """
+        super().__init__(prvs=prvs)
+
+        if len(self.prvs) != 1:
+            raise ValueError("Fourier composer can only handle homogeneous composition")
+
+        prv = self.prvs[0]
+
         if len(prv) % 2 != 0:
             raise ValueError("Can only compose evenly sized discrete PRVs")
 
         self.F = rfft(prv.pmf)
         self.domain = prv.domain
 
-    def compute_composition(self, num_compositions: int) -> DiscretePrivacyRandomVariable:
+    def compute_composition(self, num_compositions: Sequence[int]) -> DiscretePrivacyRandomVariable:
+        """Compute the composition of the PRV `num_composition` times with itself."""
+        if len(num_compositions) != 1:
+            raise ValueError("Length of `num_compositions` needs to match length of PRVs.")
```

Comment:
Would it make sense to say "not yet supported" here? Otherwise, it seems like we introduce a sequence just to ensure it contains a single element.

Reply:
That's a good point. The `Fourier` composer will only ever support compositions of a PRV with itself and we're checking for that in the constructor already.

The change in interface is mainly necessary for new composers which will allow heterogeneous composition.

Do you think it's better to change the error text here as well to indicate that Fourier only supports homogeneous composition or does it suffice to have that check in the constructor?

Reply:
Actually, I was wondering if it would be worth having an attribute for the number of prvs included in the constructor. Then the assertion here at this comment could assert on this attribute. I believe that could help for clarity in the code. What are your thoughts?

For the actual error text, it might help the user to recall the number of compositions since the `Fourier` composer could have been created some time before the `compute_composition` method is called

Reply:
To be clear, this would be an attribute of the base `Composer` and might help when moving to support heterogeneous composition later.

Reply:
I think that's a good point. But I think a new attribute is not necessary. We could just do `len(num_compositions) != len(self.prvs)` this check matches the error much closer.

Reply:
That sounds good to me. I'd also vote for the original suggestion to have the error text indicate that Fourier only supports homogeneous composition. I feel this would help debug in an ML training pipeline.